### PR TITLE
drivers: i2c: support multiple I2C addresses at runtime

### PR DESCRIPTION
**Pull request by JordanYates**
Support storing multiple reg addresses with a new Kconfig symbol, `I2C_FALLBACK_ADDRESSES`. It is relatively common for I2C devices to be available in several variants with the only difference being the I2C address. This change allows all the possible addresses to be specified. This opens up the potential for boards to populate any variant depending on component availability, without changing binaries. The alternative is defining N devices for each potential device, which comes with a much larger footprint overhead than storing a 2 byte address in ROM and a 1 byte index in RAM.

This is particularly important at the moment with component shortages everywhere. The ability to load whichever variants you
can purchase is particularly valuable.

The si7210 driver is one such example (12 variants), which has been updated to support this new feature. Below is a portion of the datasheet demonstrating the above use case.

image

This solves the problem for some simple variants of #30692.

### CODE REVIEW

**Comment by johnkjellberg** [comment]
I think this is a good start. But I hope that it can be improved on.

I2C doesn't really support discovery. You are expected to know what is where. In practice this means that if you put another chip on that address that isn't what you expect you will never know and things will just fail.

This can happen for example if you need to change multiple chips or after a couple of hw revisions.

As an example to illustrate this: I have had this exact problem where we had to replace one LED chip that was compatible, but the new had another I2C address also(call all) that had the same address as a second chip, so the second chip had to get a new address. So now chip 2 had a new address, but the original address still gives a response, but from he wrong chip.

So for both the academic/theoretical and the practical reason, I'm not a fan of this solution.

My preference (if we keep this to only solve it for I2C) would be to runtime define the address. Then the software can check the version of the board (for example by using version GPIO pins) and tell what I2C addressed to be used based on the board version.

**Comment by JordanYates** [comment]
> I2C doesn't really support discovery. You are expected to know what is where. In practice this means that if you put another chip on that address that isn't what you expect you will never know and things will just fail.

Agreed, this is not about generalized discovery, just attempting a limited subset of alternate addresses.

> This can happen for example if you need to change multiple chips or after a couple of hw revisions.

If multiple chips are changing or you have literal hardware revisions, this is not intended to be a solution. [Board revisions](https://docs.zephyrproject.org/latest/hardware/porting/board_porting.html#multiple-board-revisions ) are still the intended solution in that use case.

> So now chip 2 had a new address, but the original address still gives a response, but from he wrong chip.

Honestly I would chalk that up to bad luck, and roll the update as a new board revision.

> My preference (if we keep this to only solve it for I2C) would be to runtime define the address. Then the software can check the version of the board (for example by using version GPIO pins) and tell what I2C addressed to be used based on the board version.

Not sure what other interfaces you would want it solved for?

RE the runtime definitions, that would require moving the `i2c_dt_spec` struct out of ROM, which is a much more invasive change than this. Furthermore this change is almost sufficient to get the behaviour you want. For drivers that you want that behaviour for, expose the `spec_idx` variable through an API call (instead of attempting to auto-detect it) and provide it early in the init process from the result of the GPIO read.

**Comment by johnkjellberg** [comment]
> If multiple chips are changing or you have literal hardware revisions, this is not intended to be a solution. [Board revisions](https://docs.zephyrproject.org/latest/hardware/porting/board_porting.html#multiple-board-revisions) are still the intended solution in that use case.

Yeah. It would be really nice not to have multiple binaries for a small change like that though. It does involve extra work. So board revisions is nice in some ways, have the big drawback of multiple binaries.
 
> > My preference (if we keep this to only solve it for I2C) would be to runtime define the address. Then the software can check the version of the board (for example by using version GPIO pins) and tell what I2C addressed to be used based on the board version.
> 
> Not sure what other interfaces you would want it solved for?

I don't have a specific example of this right now. Just keeping things open. Fixing it for I2C would probably cover a majority of cases.

> RE the runtime definitions, that would require moving the `i2c_dt_spec` struct out of ROM, which is a much more invasive change than this. Furthermore this change is almost sufficient to get the behaviour you want. For drivers that you want that behaviour for, expose the `spec_idx` variable through an API call (instead of attempting to auto-detect it) and provide it early in the init process from the result of the GPIO read.

It would be nice if it was part of the API so all I2C had to implement it so it was always there if needed, without having to modify a driver and get that up streamed and so on.

Here is someone suggestion using hwinfo for a related solution. https://github.com/zephyrproject-rtos/zephyr/issues/49942#issuecomment-1237787253

File: drivers/i2c/Kconfig [hunk 1]
```diff
@@ -13,6 +13,17 @@ menuconfig I2C
 
 if I2C
 
+config I2C_FALLBACK_ADDRESSES
+	int "Enable I2C fallback addresses"
```
Comment (stephanosio) [inline drivers/i2c/Kconfig:17]: "Enable" is redundant.

Also its prompt and description are not accurate. This config is really **the number of maximum supported** I2C fallback addresses.

**Review by stephanosio — DISMISSED** [review]

**Comment by JordanYates** [comment]
> It would be nice if it was part of the API so all I2C had to implement it so it was always there if needed, without having to modify a driver and get that up streamed and so on.

I agree it would be nice, but the existing API is designed around `i2c_dt_spec` being ROM (placed in driver config structs, not data). This feature requires runtime detection, so it requires some form of mutable state, so I'm not sure if that is possible.

> Here is someone suggestion using hwinfo for a related solution. https://github.com/zephyrproject-rtos/zephyr/issues/49942#issuecomment-1237787253

That proposal is essentially a driver for reading the state of GPIOs, I'm not sure how it's relevant.

**Comment by johnkjellberg** [comment]
> > It would be nice if it was part of the API so all I2C had to implement it so it was always there if needed, without having to modify a driver and get that up streamed and so on.
> 
> I agree it would be nice, but the existing API is designed around `i2c_dt_spec` being ROM (placed in driver config structs, not data). This feature requires runtime detection, so it requires some form of mutable state, so I'm not sure if that is possible.
> 
> > Here is someone suggestion using hwinfo for a related solution. [#49942 (comment)](https://github.com/zephyrproject-rtos/zephyr/issues/49942#issuecomment-1237787253)
> 
> That proposal is essentially a driver for reading the state of GPIOs, I'm not sure how it's relevant.

To answer both. By having the version implemented in a standard way Zephyr can know about it and therefore will be able to select I2C address based on it or forward it to the driver. The driver then needs to get that information in the init function on one way or another. Or Zephyr can have helper functions that uses that driver to get the correct configuration, that the I2C driver can use. The figured out valid I2C address will have to live in RAM no matter how it is done, so I don't really see the ROM/RAM conflict here.

The key is just that how to read those pins/the hw version needs to be abstracted away in a standard way for the drivers and/or Zephyr to be able to use that information.

**Comment by JordanYates** [comment]
> The key is just that how to read those pins/the hw version needs to be abstracted away in a standard way for the drivers and/or Zephyr to be able to use that information.

I disagree that an API to return a hardware version is the "key" to the features described. It is definitely required, but it is also the first and easiest step. I would love to be proven wrong, but I don't see an easy path toward what you describe, maybe just because it is lacking in detail.

**Comment by johnkjellberg** [comment]
> > The key is just that how to read those pins/the hw version needs to be abstracted away in a standard way for the drivers and/or Zephyr to be able to use that information.
> 
> I disagree that an API to return a hardware version is the "key" to the features described. It is definitely required, but it is also the first and easiest step. I would love to be proven wrong, but I don't see an easy path toward what you describe, maybe just because it is lacking in detail.

Sorry. I meant that that was the key reason that I linked to that. It is a small, but important part, to get it working. It is not the key to the whole solution.

However, I just realized that this problem isn't that different from dynamic pin control in some ways. https://docs.zephyrproject.org/latest/hardware/pinctrl/index.html#dynamic-pin-control . It could perhaps be modeled more like that. So that custom code will be executed to selected configuration based on whatever criteria you want, including probing. Pin control is something that was very static, that was made dynamic with that feature.

However, I am a bit to new and unfamiliar with the internals of zephyr in this case to provide an exact solution. But I haven't seen anything yet that that I feel would make it (practically) impossible. And I do believe that many people would find it useful and it would add lot of value to have it cover as many use cases as possible.

**Comment by carlescufi** [comment]
@teburd can you please take a look?

**Review by teburd — CHANGES_REQUESTED** [review]
I'd really like to know what others think of this.

I don't hate it, but I don't love it either.

I guess I'm wondering why a driver can't do this itself by probing the bus for all the possible addresses at init time and choose then.

There's no requirement to use i2c_dt_spec, that's there for convenience. If a driver can take advantage of multiple addresses that seems like a driver specific thing.


E.g.
``` c
struct my_driver_data {
  uint16_t i2c_addr;
}

struct my_driver_config {
  struct device *bus
  uint16_t *probe_addrs;
  uint32_t probe_addrs_len;
}

int my_driver_init(const struct device *dev) {
  for(int i = 0; i < cfg->probe_addrs_len; i++) {
    if(my_driver_probe(dev, cfg->probe_addrs[i])) {
      break;
    }
  }
}
```

Then use the normal i2c_read/write/etc functions.

I'm not sure there's a big benefit, in my mind, to storing an array of addresses in rom and an index into that array in ram vs just storing the usable address in ram and using the normal i2c functions.

**Review by henrikbrixandersen — APPROVED** [review]

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]
Blocking as I do not agree with the design choices made here.

**Comment by JordanYates** [comment]
> I don't hate it, but I don't love it either.

I am on the same page here. I wrote this to support a real use-case I had an submitted it upstream in case it was useful.

> I guess I'm wondering why a driver can't do this itself by probing the bus for all the possible addresses at init time and choose then.

Of course they could, but as soon as you have more than 2 drivers doing this, looping over the `reg = <>` fields may as well be in a common helper function.

> There's no requirement to use i2c_dt_spec, that's there for convenience. If a driver can take advantage of multiple addresses that seems like a driver specific thing.

A lot of things in Zephyr exist for convenience. If this is a problem that could be solved/made easier for multiple drivers in one place I think that is useful.

**Comment by JordanYates** [comment]
> Blocking as I do not agree with the design choices made here.

Is it the feature itself that you disagree with or the implementation?

**Pull request closed without merging**